Add 90CW, 90CCW and 180  rotations

f-fl0 opened the pull request:
This PR adds a rotate function to apply 90 degrees CW, 90 degrees CCW or 180 degrees rotation to the captured frames.

f-fl0 committed:
Function to rotate images by 90CW, 90CCW or 180

- Add rotate_code parameter.
- Add function to rotate images by 90CW, 90CCW or 180 via array
manipulation.

f-fl0 committed:
Update camera info message

- Adjust camera matrix.
- Adjust ROI.

at-wat commented on nodes/usb_cam_node.cpp:
```diff
@@ -48,6 +48,10 @@
 
+const static int ROTATE_90_CW = 0;
+const static int ROTATE_90_CCW = 1;
+const static int ROTATE_180 = 2;
```
They may be better to be `enum class`

at-wat commented on nodes/usb_cam_node.cpp:
```diff
@@ -162,6 +169,8 @@ class UsbCamNode
     node_.param("gamma_max", gamma_max_, 200);
     node_.param("gamma_min", gamma_min_, 50);
 
+    node_.param("rotate_code", rotate_code_, -1);
```
I think it's better to make the values readable.
e.g.
- make it `int` and check the value is one of `90`, `-90`, `180`
- make it string and allow `90CW`, `90CCW`, `180`

at-wat commented on nodes/usb_cam_node.cpp:
```diff
@@ -334,12 +343,123 @@ class UsbCamNode
+        if (rotate_code_ == ROTATE_90_CW)
+        {
+          // update cx and cy
+          double tmp_cx = ci->K[2];
```
const

at-wat commented on nodes/usb_cam_node.cpp:
```diff
@@ -334,12 +343,123 @@ class UsbCamNode
+          // update ROI
+          if (ci->roi.width != 0 && ci->roi.height != 0)
+          {
+            double tmp_x_offset = ci->roi.x_offset;
```
const

at-wat commented on nodes/usb_cam_node.cpp:
```diff
@@ -334,12 +343,123 @@ class UsbCamNode
+        if (rotate_code_ == ROTATE_90_CCW)
+        {
+          // update cx and cy
+          double tmp_cx = ci->K[2];
```
const

at-wat commented on nodes/usb_cam_node.cpp:
```diff
@@ -334,12 +343,123 @@ class UsbCamNode
+          // update ROI
+          if (ci->roi.width != 0 && ci->roi.height != 0)
+          {
+            double tmp_x_offset = ci->roi.x_offset;
```
const

at-wat commented on nodes/usb_cam_node.cpp:
```diff
@@ -334,12 +343,123 @@ class UsbCamNode
 
-    // publish the image
-    image_pub_.publish(*img_ptr, *ci);
+    if (rotate_code_ == ROTATE_90_CW || rotate_code_ == ROTATE_90_CCW || rotate_code_ == ROTATE_180)
```
It would be better to make the header update logic a separate function and add a simple test.

at-wat commented on nodes/usb_cam_node.cpp:
```diff
@@ -334,12 +343,123 @@ class UsbCamNode
     return true;
   }
 
+  void rotate(uint8_t *src, uint8_t *dst, const int row, const int col, const int ch, int rotate_code)
```
```suggestion
  void rotate(uint8_t *src, uint8_t *dst, const int row, const int col, const int ch, const int rotate_code)
```

at-wat commented on nodes/usb_cam_node.cpp:
```diff
@@ -334,12 +343,123 @@ class UsbCamNode
+          for (int c = 0; c < ch; c++)
+          {
+            // rotate 90 degrees CW
+            if (rotate_code == ROTATE_90_CW)
```
I guess `if` inside these loops affects the performance optimization.

f-fl0 committed:
Address PR comments

- Use enum instead of constants.
- Use const when appropriate.
- No if whithin for loops.
- Separate function to update camera info.

f-fl0 commented:
@at-wat I addressed most of your comments in 8012a5830d030875306640ad3411c9f62f6b5946 (only test is mssing).

f-fl0 committed:
Add unittest for rotate function

- Separate the rotate related function to a different file.
- Add unittests about the camera info update function.

f-fl0 committed:
Remove extra line.

f-fl0 committed:
Remove unnecessary print.

f-fl0 commented:
@at-wat PTAL

at-wat commented on include/usb_cam/rotate.h:
```diff
@@ -0,0 +1,135 @@
+
```
Please add license info (BSD https://github.com/seqsense/usb_cam/blob/sq-master/LICENSE copyright by SEQSENSE)

f-fl0 replied:
Done in 8388e4148800e642fa919a0fe1ee945498747428

at-wat commented on include/usb_cam/rotate.h:
```diff
@@ -0,0 +1,135 @@
+
+#ifndef USB_CAM_ROTATE_H
+#define USB_CAM_ROTATE_H
+
+#include "sensor_msgs/CameraInfo.h"
+
+namespace usb_cam {
+
+enum RotateCode { ROTATE_NONE, ROTATE_90_CW, ROTATE_90_CCW, ROTATE_180 };
+
+void update_camera_info(sensor_msgs::CameraInfoPtr ci, const RotateCode rotate_code)
+{
+  if (rotate_code == ROTATE_NONE)
+    return;
+
+  if (rotate_code == ROTATE_180)
+  {
+    // update cx and cy
+    ci->K[2] = ci->width - ci->K[2];
+    ci->K[3 + 2] = ci->height - ci->K[3 + 2];
+    ci->P[2] = ci->width - ci->P[2];
+    ci->P[4 + 2] = ci->height - ci->P[4 + 2];
+
+    // update ROI
+    if (ci->roi.width != 0 && ci->roi.height != 0)
+    {
+      ci->roi.x_offset = ci->width - ci->roi.x_offset;
+      ci->roi.y_offset = ci->height - ci->roi.y_offset;
+    }
+  }
+  else
+  {
```
```suggestion
    }
    return;
  }
```

f-fl0 replied:
I'm not sure I understand the suggestion here.

at-wat replied:
Currently this block is like:
```cpp
void update_camera_info(sensor_msgs::CameraInfoPtr ci, const RotateCode rotate_code)
{
  ...
  if (rotate_code == ROTATE_180)
  {
    ...
  }
  else
  {
    ...
  }
}
```

Following would be simpler:
```cpp
void update_camera_info(sensor_msgs::CameraInfoPtr ci, const RotateCode rotate_code)
{
  ...
  if (rotate_code == ROTATE_180)
  {
    ...
    return;
  }
  ...
}
```

f-fl0 replied:
Done in c08d095534181f1a00372bef60aebe0ee56e5327

at-wat commented on include/usb_cam/rotate.h:
```diff
@@ -0,0 +1,135 @@
+  }
+}
+
+void rotate(uint8_t *src, uint8_t *dst, const int row, const int col, const int ch, const RotateCode rotate_code)
```
```suggestion
void rotate(const uint8_t *src, uint8_t *dst, const int row, const int col, const int ch, const RotateCode rotate_code)
```

f-fl0 replied:
Done in 8388e4148800e642fa919a0fe1ee945498747428

at-wat commented on test/test_rotate.cpp:
```diff
@@ -0,0 +1,119 @@
+#include <gtest/gtest.h>
```
copyright also to this file

f-fl0 replied:
Done in 8388e4148800e642fa919a0fe1ee945498747428

f-fl0 committed:
Address PR comments

- Add license information.
- Add const.

f-fl0 committed:
Address PR comments

- Simplify if/else blocks.

at-wat reviewed (approved):
LGTM
